src/Model/Aggregate.php

```diff
@@ -0,0 +1,316 @@
+
+use Atk4\Data\Exception;
+use Atk4\Data\Field;
+use Atk4\Data\FieldSqlExpression;
+use Atk4\Data\Model;
+use Atk4\Data\Persistence;
+use Atk4\Data\Reference;
+use Atk4\Dsql\Expression;
+use Atk4\Dsql\Query;
+
+/**
+ * Aggregate model allows you to query using "group by" clause on your existing model.
+ * It's quite simple to set up.
+ *
+ * $aggregate = new Aggregate($mymodel);
+ * $aggregate->groupBy(['first','last'], ['salary'=>'sum([])'];
+ *
+ * your resulting model will have 3 fields:
+ *  first, last, salary
+ *
+ * but when querying it will use the original model to calculate the query, then add grouping and aggregates.
+ *
+ * If you wish you can add more fields, which will be passed through:
+ * $aggregate->addField('middle');
+ *
+ * If this field exist in the original model it will be added and you'll get exception otherwise. Finally you are
+ * permitted to add expressions.
+ *
+ * The base model must not be Union model or another Aggregate model, however it's possible to use Aggregate model as nestedModel inside Union model.
+ * Union model implements identical grouping rule on its own.
+ *
+ * You can also pass seed (for example field type) when aggregating:
+ * $aggregate->groupBy(['first','last'], ['salary' => ['sum([])', 'type'=>'money']];
+ *
+ * @property \Atk4\Data\Persistence\Sql $persistence
+ *
+ * @method Expression expr($expr, array $args = []) forwards to Persistence\Sql::expr using $this as model
+ */
+class Aggregate extends Model
+{
+    /** @const string */
+    public const HOOK_INIT_SELECT_QUERY = self::class . '@initSelectQuery';
+
+    /** @var array */
+    protected $systemFields = [];
+
+    /** @var Model */
+    public $baseModel;
+
+    /**
+     * Aggregate model should always be read-only.
+     *
+     * @var bool
+     */
+    public $read_only = true;
+
+    /**
+     * Aggregate does not have ID field.
+     *
+     * @var string
+     */
+    public $id_field;
+
+    /** @var array */
+    public $group = [];
+
+    /** @var array */
+    public $aggregate = [];
+
+    /**
+     * Constructor.
+     */
+    public function __construct(Model $baseModel, array $defaults = [])
+    {
+        if (!$baseModel->persistence instanceof Persistence\Sql) {
+            throw new Exception('Base model must have Sql persistence to use grouping');
+        }
+
+        $this->baseModel = clone $baseModel;
+        $this->table = $baseModel->table;
+
+        parent::__construct($baseModel->persistence, $defaults);
+
+        // always use table prefixes for this model
+        $this->persistence_data['use_table_prefixes'] = true;
+    }
+
+    /**
+     * Specify a single field or array of fields on which we will group model.
+     *
+     * @param array $fields    Array of field names
+     * @param array $aggregate Array of aggregate mapping
+     *
+     * @return $this
+     */
+    public function groupBy(array $fields, array $aggregate = []): Model
+    {
+        $this->group = $fields;
```

Comment:
I think we should allow to call this method multiple times.

`$aggregate` param - what is the advantage of it instead of `addExpression` or this nodel or simillar method?

Reply:
> I think we should allow to call this method multiple times.

I did modify it already

> $aggregate param - what is the advantage of it instead of addExpression or this nodel or simillar method?

I played a bit to use addExpression natively but it mixes things up as aggregate expressions use base model fields. Then also it messes up with base model addExpression and you cannot do `$model->addExpression('xx')->groupBy(['yy'])` but need to do `$model->groupBy` first to create the aggregate. So multiple issues in this approach.

Reply:
I am ok with this for now, as related only with this method...

Reply:
fixed